src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs

```diff
@@ -387,36 +387,18 @@ public void SetStateMachine(IAsyncStateMachine stateMachine)
         {
             IAsyncStateMachineBox box = GetStateMachineBox(ref stateMachine);
 
-            // TODO https://github.com/dotnet/coreclr/issues/12877:
-            // Once the JIT is able to recognize "awaiter is ITaskAwaiter" and "awaiter is IConfiguredTaskAwaiter",
-            // use those in order to a) consolidate a lot of this code, and b) handle all Task/Task<T> and not just
-            // the few types special-cased here.  For now, handle common {Configured}TaskAwaiter.  Having the types
-            // explicitly listed here allows the JIT to generate the best code for them; otherwise we'll fall through
-            // to the later workaround.
-            if (typeof(TAwaiter) == typeof(TaskAwaiter) ||
-                typeof(TAwaiter) == typeof(TaskAwaiter<object>) ||
-                typeof(TAwaiter) == typeof(TaskAwaiter<string>) ||
-                typeof(TAwaiter) == typeof(TaskAwaiter<byte[]>) ||
-                typeof(TAwaiter) == typeof(TaskAwaiter<bool>) ||
-                typeof(TAwaiter) == typeof(TaskAwaiter<byte>) ||
-                typeof(TAwaiter) == typeof(TaskAwaiter<int>) ||
-                typeof(TAwaiter) == typeof(TaskAwaiter<long>))
+            if (null != (object)default(TAwaiter))
```

Comment:
Any benefit in extending this guard to also cover the branches for ValueTaskAwaiter and ConfiguredValueTaskAwaiter?  Those are both structs as well.  Or maybe it doesn't matter because the JIT is already good at the explicit type comparisons used there?

Reply:
Actually, I don't think as written this is correct.  This will effectively be true for any non-nullable struct, right?  In which case, this change causes the below ValueTaskAwaiter optimizations to be skipped.

Reply:
Yeah, I think you're right. I should inline the value class checks instead, something like:
```C#
            if ((null != (object)default(TAwaiter)) && (awaiter is ITaskAwaiter))
            {
                ref TaskAwaiter ta = ref Unsafe.As<TAwaiter, TaskAwaiter>(ref awaiter); // relies on TaskAwaiter/TaskAwaiter<T> having the same layout
                TaskAwaiter.UnsafeOnCompletedInternal(ta.m_task, box, continueOnCapturedContext: true);
            }
            else if ((null != (object)default(TAwaiter)) && (awaiter is IConfiguredTaskAwaiter))
            {
                ref ConfiguredTaskAwaitable.ConfiguredTaskAwaiter ta = ref Unsafe.As<TAwaiter, ConfiguredTaskAwaitable.ConfiguredTaskAwaiter>(ref awaiter);
                TaskAwaiter.UnsafeOnCompletedInternal(ta.m_task, box, ta.m_continueOnCapturedContext);
            }
```